Run SQL on larger same-version group (#19026)
hazelcast/hazelcast: site/docs/design-docs/021-sql-rolling-upgrade.md

engine to be available during this time. We'll add a new
`JobConfig.blockShutdown` option. The SQL engine will enable this option
for batch jobs, but not for streaming jobs.

Jobs submitted during this time will not use the member in the shutdown
process as a coordinator. Other members can handle new jobs, but they
will not use the members being shut down. To do this, the existing
`NotifyMemberShutdownOperation` will be broadcast to all members
(currently it's sent just to the master).

The shutdown procedure will block until all jobs with this option
complete.

Fault-tolerant jobs will be shut down immediately with a snapshot,
regardless of this option. Other jobs (non-fault-tolerant ones without
this option enabled) will be cancelled (light jobs) or restarted (normal
jobs).

It can happen that the member will never shut down if the user submits a
streaming job with the `blockShutdown` option. For this scenario there
is a hard-coded timeout of 10 seconds after which the member proceeds
with the shutdown without waiting for all jobs to terminate. In this
case the administrator must be able to see that job and cancel it

> manually, using either Management center or Java API. Another option is

Reviewer: So, is it a timeout or manual cancellation?

Author: Both. I will finalize this as a part of the 2nd PR.

### Understandable error messages for unsupported scenarios.

Unsupported features must fail with a clear error message. E.g. a normal
job must fail with appropriate error message after a member with a newer
version is added, even in a fault-tolerant job.

### Changes to the SQL engine

For non-smart client, the `SqlExecute` client operation will route to a
random member - it might not be from the larger same-version group. To
address this, we need to add member-to-member variants of `SqlExecute`,
`SqlFetch` and `SqlCancel` operations so that a query submitted from a
non-smart client can be redirected to execute on a member from the
larger same-version subset. This wil also address the current limitation
that SQL queries can't be submitted from lite members.

### LoadBalancer changes

`com.hazelcast.client.LoadBalancer` is public API. Has 3
implementations: `StaticLB`, `RoundRobinLB` and `RandomLB`. The
`AbstractLoadBalancer` base implementation is also public API.

In 4.2 `LoadBalancer.nextDataMember()` method was added that throws
`UnsupportedOperationException` in the default implementation. In 4.2,
the SQL API required that the query was submitted to a data member. If

> the user had he's own LB implementation, SQL won't work from clients at

Reviewer: ```suggestion
the user had his own LB implementation, SQL would not work from clients at
```

directly to just one member which needs to forward the request if it's
not from the SQL group.

### Handling of client connections

Queries submitted from a client should be cancelled when that client
disconnects. Before, the client operation was always handled by the
coordinating member, but now it could be forwarded to a different
member. To handle this, we need to add two actions:

1. the member that received the client message, will monitor the client,
and if it disconnects, it will send the `SqlCloseOperation` to the
coordinator to cancel the job.
   
2. the job coordinator will cancel the job if the submitting member
leaves

### Client security

If security in Hazelcast Enterprise is enabled, for queries submitted
from a client we must check permissions. However, the member handling
the client request isn't able to check them until it parses and
validates the query. And since the query can be forwarded to a different
member, we need to forward the authentication information to the
coordinating member. Therefore we add `Collection<Principal>` to the
`SqlExecuteOperation`. If this collection is not null, the recipient

> will check if som of the principals has access to the objects referenced

Reviewer: ```suggestion
will check if some of the principals has access to the objects referenced
```
